content/blogposts/2024/improving-cody-autocomplete-faster-smarter.md

```diff
@@ -0,0 +1,132 @@
+
+Unlike traditional caching that might only store single-line completions, Hot Streak proactively caches multiple lines of potential completions. This allows for faster suggestion delivery of multi-line code snippets without the need to make additional server requests.
+
+The implementation of Hot Streak contributed significantly to Cody's performance improvements. While not explicitly quantified in isolation, it was part of the improvements that led to:
+
+1. A reduction in P75 latency by approximately 110ms across all line-count groups.
+2. A 24% increase in the number of suggested characters.
+3. A 10% increase in the total number of accepted characters, indicating that Cody wrote more code for users.
+4. Approximately 35% increase in completions served from cache (in combination with Smart Throttle prefetching).
+
+<Figure 
+    src="https://storage.googleapis.com/sourcegraph-assets/blog/faster-autocomplete/autocomplete-hotstreak.png"
+    caption="Faster completions served from cache"
+/>
+
+### Implementing Smart Throttle
+
+Smart Throttle is a client-side optimization technique implemented in Cody to improve autocomplete responsiveness. It works by intelligently reducing the debounce interval between user input and completion requests.
+
+Unlike traditional fixed debounce timers, Smart Throttle dynamically adjusts its timing based on typing patterns and context. This allows for faster suggestion delivery without overwhelming the server with excessive requests.
+
+The implementation of Smart Throttle contributed significantly to Cody's performance improvements. It helped reduce overall client-side latency by about 110ms across all line-count groups. Additionally, it played a role in **increasing the number of suggested characters by 24% and boosting cache utilization by 35%**.
+
+## Smarter Suggestions: 58% More Accepted Code
+
+Not only is Cody faster, but it's also getting better at suggesting code you actually want to use. The DeepSeek v2 model has significantly improved the quality of suggestions. 
+
+We’ve seen a **58% increase** in the number of accepted characters per user. This means that the suggestions provided by Cody are more likely to be relevant and accepted. In practical terms, if you're writing a 1000-line program, Cody is now potentially writing 580 of those lines for you, up from about 367 before.
+
+- Suggested characters per user increased from 1,600 to 2,500.
+- Accepted characters per user increased from 600 to 950.
+
+Here’s our analytics tracking suggested characters and accepted characters:
+
+<Figure 
+    src="https://storage.googleapis.com/sourcegraph-assets/blog/faster-autocomplete/autocomplete-characters.png"
+    caption="Suggested and accepted characters per user"
+/>
+
+## Better Multi-Line Completions
+
+Our A/B testing showed significant improvements for multi-line completions, thanks to DeepSeek v2's enhanced ability to understand context:
```

Comment:
Should this be a colon, since there's a colon on the line right below?

```suggestion
Our A/B testing showed significant improvements for multi-line completions, thanks to DeepSeek-V2's enhanced ability to understand context.
```